lib/src/fix.rs

```diff
@@ -0,0 +1,319 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! API for transforming file content, for example to apply formatting, and
+//! propagate those changes across revisions.
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::mpsc::channel;
+
+use futures::StreamExt;
+use itertools::Itertools;
+use jj_lib::backend::BackendError;
+use jj_lib::backend::CommitId;
+use jj_lib::backend::FileId;
+use jj_lib::backend::TreeValue;
+use jj_lib::matchers::Matcher;
+use jj_lib::merged_tree::MergedTree;
+use jj_lib::merged_tree::MergedTreeBuilder;
+use jj_lib::merged_tree::TreeDiffEntry;
+use jj_lib::repo::MutableRepo;
+use jj_lib::repo::Repo;
+use jj_lib::repo_path::RepoPathBuf;
+use jj_lib::revset::RevsetExpression;
+use jj_lib::revset::RevsetIteratorExt;
+use jj_lib::store::Store;
+use jj_lib::tree::Tree;
+use pollster::FutureExt;
+use rayon::iter::IntoParallelIterator;
+use rayon::prelude::ParallelIterator;
+
+use crate::revset::RevsetEvaluationError;
+
+/// Represents a file whose content may be transformed by a FileFixer.
+// TODO: Add the set of changed line/byte ranges, so those can be passed into code formatters via
+// flags. This will help avoid introducing unrelated changes when working on code with out of date
+// formatting.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct FileToFix {
+    /// Unique identifier for the file content.
+    pub file_id: FileId,
+
+    /// The path is provided to allow the FileFixer to potentially:
+    ///  - Choose different behaviors for different file names, extensions, etc.
+    ///  - Update parts of the file's content that should be derived from the
+    ///    file's path.
+    pub repo_path: RepoPathBuf,
+}
+
+/// Error fixing files.
+#[derive(Debug, thiserror::Error)]
+pub enum FixError {
+    /// Error while contacting the Backend.
+    #[error(transparent)]
+    Backend(#[from] BackendError),
+    /// Error resolving commit ancestry.
+    #[error(transparent)]
+    RevsetEvaluation(#[from] RevsetEvaluationError),
+    /// Error occurred while reading/writing file content.
+    #[error(transparent)]
+    IO(#[from] std::io::Error),
+    /// Error occurred while processing the file content.
+    #[error(transparent)]
+    FixContent(Box<dyn std::error::Error + Send + Sync>),
+}
+
+/// Fixes a set of files.
+///
+/// Fixing a file is implementation dependent. For example it may format source
+/// code using a code formatter.
+pub trait FileFixer {
```

Comment:
Can you split patches into trivial code move and design changes (such as introduction of `FileFixer`)? It's also nice if the commit description includes why `FileFixer` trait is added, and what kind of implementations are supposed to be added, etc.

Reply:
That was not easy to do because fix in the CLI has logic that we do not want to move which is interspersed (both as callee and caller) with logic we do want in lib. Having said that, I am open to push back.

Reply:
I think some part could be left in CLI as callback functions (and/or moved to jj-lib incrementally)?

I sometimes do refactoring before moving functions to jj-lib, but I agree it would be waste of time because you've already implemented differently.

Reply:
Ok, let me know if there is anything I could do here. Once jj has copy/move support this will become easier.